WorkflowTemplate CRD

dtaniwaki opened the pull request:
I implemented `WorkflowTemplate` CRD which holds template data outside of `Workflow` which was discussed in #927.

You can refer to another template from any template by either local template name of `template` or template resource reference of `templateRef`. When the referenced template is loaded, template level arguments of `arguments` are applied.

Here's the example which describes the feature well.

```
apiVersion: argoproj.io/v1alpha1
kind: WorkflowTemplate
metadata:
  name: workflow-template-whalesay-template
spec:
  templates:
  - name: whalesay-template
    inputs:
      parameters:
      - name: message
    container:
      image: docker/whalesay
      command: [cowsay]
      args: ["{{inputs.parameters.message}}"]
---
apiVersion: argoproj.io/v1alpha1
kind: WorkflowTemplate
metadata:
  name: workflow-template-nested-template
spec:
  templates:
  - name: whalesay-inner-template
    templateRef:
      name: workflow-template-whalesay-template
      template: whalesay-template
    inputs:
      parameters:
      - name: message
    arguments:
      parameters:
      - name: message
        value: "{{inputs.parameters.message}} hello"
  - name: whalesay-template
    template: whalesay-inner-template
    inputs:
      parameters:
      - name: message
    arguments:
      parameters:
      - name: message
        value: "{{inputs.parameters.message}} hello"
---
apiVersion: argoproj.io/v1alpha1
kind: Workflow
metadata:
  generateName: workflow-template-nested-
spec:
  entrypoint: whalesay
  templates:
  - name: whalesay
    templateRef:
      name: workflow-template-nested-template
      template: whalesay-template
    arguments:
      parameters:
      - name: message
        value: "hello"
```

Here's the guide to read the example above.
- It starts from the entry point `whalesay` of the workflow `workflow-template-nested-xxxx`
- The template refers to the template `whalesay-template` in the workflow template `workflow-template-nested-template`.
- The referred template refers to the template `whalesay-inner-template` of the context of the workflow template `workflow-template-nested-template`.
- The referred template refers to the template `whalesay-template` of the context of the workflow template `workflow-template-whalesay-template`.
- The template is a concrete template so returns this template.

After resolving input parameters of referred templates by arguments of referring templates at each step, this workflow outputs `hello hello hello` in the log.

Workflow templates must be submitted with `argo template submit` command before workflows using them are submitted unless you set the `runtimeResolution` option.

dtaniwaki commented:
I'm still adding e2e tests and unit tests as the PR is quite huge.

dtaniwaki commented:
I'll also check if ArgoUI works with this template well.

dtaniwaki commented:
Here's the list of tasks I'm working on for the PR.
- Add e2e and unit tests
- Use shared indexer to get templates
- Resolve templates on validate
- Make sure ArgoUI still works with templates

sarabala1979 commented:
Do you check templateRef is already load or not during workflow submit?

dtaniwaki commented:
Yes! I will validate resolved templates when a workflow is submitted.

dtaniwaki commented:
I added tests and template resolution on valdate. I didn't implement shared informer support to reduce workflow template get requests because it takes more time to implement it but the current implementation may be changed after reviews. I'll make another PR for it after this PR gets merged.

Workflows with WorkflowTemplate can be displayed in ArgoUI although there's no way to check the template contents on UI. I'll add template resolution on UI after this PR gets merged too.

dtaniwaki commented:
@jessesuen @alexmt Could you review the PR?

xlucas commented:
Thanks for your work! 🎉 
I'm trying to run the example from the PR description. For now I've encountered two issues:

- I can't submit the full example to argo, I have to first submit workflowTemplate CRDs first then submit the workflow, otherwise I get:
```
2019/04/23 09:53:48 Failed to submit workflow: spec.entrypoint workflow template workflow-template-nested-template not found
```
- If I create the WorkflowTemplate CRDs prior to submitting the workflow, the execution fails afterwards and `argo get` shows:

```
argo get workflow-template-nested-pgmcs
Name:                workflow-template-nested-pgmcs
Namespace:           default
ServiceAccount:      default
Status:              Failed
Message:             invalid spec: template type unspecified. choose one of: container, steps, script, resource, dag, suspend
Created:             Tue Apr 23 10:01:51 +0000 (10 seconds ago)
Started:             Tue Apr 23 10:01:51 +0000 (10 seconds ago)
Finished:            Tue Apr 23 10:01:51 +0000 (10 seconds ago)
Duration:            0 seconds
```

I've tried with both `v2.3.0-rc1` and a local build from the latest version of this branch.

xlucas commented on workflow/validate/validate.go:
```diff
@@ -163,10 +198,27 @@ func (ctx *wfValidationCtx) validateTemplate(tmpl *wfv1.Template, args wfv1.Argu
 func validateTemplateType(tmpl *wfv1.Template) error {
 	numTypes := 0
-	for _, tmplType := range []interface{}{tmpl.Container, tmpl.Steps, tmpl.Script, tmpl.Resource, tmpl.DAG, tmpl.Suspend} {
+	for _, tmplType := range []interface{}{tmpl.TemplateRef, tmpl.Container, tmpl.Steps, tmpl.Script, tmpl.Resource, tmpl.DAG, tmpl.Suspend} {
```
The error message for an unspecified template type should include `templateRef`, for now it only mentions `container, steps, script, resource, dag, suspend`.

dtaniwaki commented:
@xlucas Thank you for checking the function! Are you using minikube? If so, you need to build the image with minikue env.

xlucas commented:
@dtaniwaki I'm using hyperkube but that was a pretty similar issue with old images being pulled. This looks fine now!

dtaniwaki commented:
@xlucas I actually want to have feedback about the CLI for the workflow template feature. I wanted to handle workflow templates in the same sub command for workflows, but I was not sure the best way to print workflow templates with workflows. Maybe, just print two tables for workflows and workflow templates? If we can handle workflow templates like this, we don’t need extra su command like `argo template submit`.

xlucas commented:
@dtaniwaki I do believe this a good think to have separate subcommands for separate objects.

Even though it may be confusing for newcomers not to be able to submit an all-in-one *"kubernetes-style"* manifest, I think it's not a problem as long as it's properly documented. Otherwise, operating templates would be tightly coupled to submitting specific workflows whilst they serve the opposite purpose. WDYT?

dtaniwaki commented:
I realized I need to use the default service account of workflow’s namespace to refer workflow templates and maybe need to add something like workflow template service account option. I will implement it.

hnykda commented:
Hey. This looks great - is it planned to be merged soon?

dtaniwaki commented:
@hnykda I was busy the past month and I still need to consider the accessibility of the WorkflowTemplate CRD. I'll tackle it soon!

frankfarzan commented:
@dtaniwaki This is a really useful feature to make templates reusable! Would love to see it merged and try it out.

dtaniwaki commented:
> I realized I need to use the default service account of workflow’s namespace to refer workflow templates, otherwise, anyone can access to any workflow templates. Maybe we need to add something like workflow template service account option. I will implement it.

After thinking of this issue a while, I concluded that we don't have to care the accessibility of workflow templates as long as the templates exist in the same namespace of a workflow which refers to them like the same way as other resources like ConfigMap are handled.

dtaniwaki commented:
I refactored the code a little bit and I think it's pretty much of the basic function of this feature.

Here's the output of argo cli when I submitted workflows w/ templates.

```
$ ./dist/argo template submit -n argo examples/workflow-template/*.yaml
WARN[0000] Workflow is not a workflow template
WARN[0000] Workflow is not a workflow template
WARN[0000] Workflow is not a workflow template
WARN[0000] Workflow is not a workflow template
WARN[0000] Workflow is not a workflow template
WARN[0000] Workflow is not a workflow template
Name:                workflow-template-whalesay-template
Namespace:           argo
Created:             Mon Jul 15 01:08:13 +0900 (1 second ago)
Name:                workflow-template-nested-template
Namespace:           argo
Created:             Mon Jul 15 01:08:14 +0900 (now)
Name:                workflow-template-random-fail-template
Namespace:           argo
Created:             Mon Jul 15 01:08:14 +0900 (now)
$ ./dist/argo submit -n argo examples/workflow-template/*.yaml
WARN[0000] WorkflowTemplate is not a workflow
WARN[0000] WorkflowTemplate is not a workflow
WARN[0000] WorkflowTemplate is not a workflow
Name:                workflow-template-dag-diamond-4q2rh
Namespace:           argo
ServiceAccount:      default
Status:              Pending
Created:             Mon Jul 15 01:08:17 +0900 (now)
Name:                workflow-template-hello-world-ptcvr
Namespace:           argo
ServiceAccount:      default
Status:              Pending
Created:             Mon Jul 15 01:08:18 +0900 (now)
Name:                hello-world-zk7qr
Namespace:           argo
ServiceAccount:      default
Status:              Pending
Created:             Mon Jul 15 01:08:18 +0900 (now)
Name:                workflow-template-nested-2mdjk
Namespace:           argo
ServiceAccount:      default
Status:              Pending
Created:             Mon Jul 15 01:08:20 +0900 (now)
Name:                workflow-template-retry-with-steps-6542b
Namespace:           argo
ServiceAccount:      default
Status:              Pending
Created:             Mon Jul 15 01:08:21 +0900 (now)
Name:                workflow-template-steps-g6llt
Namespace:           argo
ServiceAccount:      default
Status:              Pending
Created:             Mon Jul 15 01:08:23 +0900 (now)
$ argo list -n argo
NAME                                       STATUS      AGE   DURATION   PRIORITY
workflow-template-dag-diamond-4q2rh        Succeeded   1m    1m         0
workflow-template-retry-with-steps-6542b   Succeeded   1m    56s        0
workflow-template-steps-g6llt              Succeeded   1m    50s        0
workflow-template-nested-2mdjk             Succeeded   1m    15s        0
workflow-template-local-ref-zk7qr                          Succeeded   1m    14s        0
workflow-template-hello-world-ptcvr        Succeeded   1m    11s        0
$ ./dist/argo get -n argo workflow-template-retry-with-steps-6542b
Name:                workflow-template-retry-with-steps-6542b
Namespace:           argo
ServiceAccount:      default
Status:              Succeeded
Created:             Mon Jul 15 01:08:21 +0900 (2 minutes ago)
Started:             Mon Jul 15 01:08:21 +0900 (2 minutes ago)
Finished:            Mon Jul 15 01:09:17 +0900 (1 minute ago)
Duration:            56 seconds

STEP                                                                               PODNAME                                              DURATION  MESSAGE
 ✔ workflow-template-retry-with-steps-6542b (retry-with-steps)                                                      
 ├---✔ hello1 (workflow-template-random-fail-template/random-fail-template)                                         
 |   ├-✖ hello1(0) (workflow-template-random-fail-template/random-fail-template)   workflow-template-retry-with-steps-6542b-3456092177  33s       failed with exit code 1
 |   └-✔ hello1(1) (workflow-template-random-fail-template/random-fail-template)   workflow-template-retry-with-steps-6542b-4060233556  4s
 └-·-✔ hello2a(0) (workflow-template-random-fail-template/random-fail-template)    workflow-template-retry-with-steps-6542b-3526014558  8s
   └-✔ hello2b (workflow-template-random-fail-template/random-fail-template)                                        
     ├-✖ hello2b(0) (workflow-template-random-fail-template/random-fail-template)  workflow-template-retry-with-steps-6542b-3678971621  8s        failed with exit code 1
     ├-✖ hello2b(1) (workflow-template-random-fail-template/random-fail-template)  workflow-template-retry-with-steps-6542b-3477787288  3s        failed with exit code 1
     └-✔ hello2b(2) (workflow-template-random-fail-template/random-fail-template)  workflow-template-retry-with-steps-6542b-1329810771  3s
```

dtaniwaki commented:
I updated the steps and DAG example so we can check how nested steps/DAG work.

```
$ ./dist/argo get -n argo workflow-template-steps-dlqqb
Name:                workflow-template-steps-dlqqb
Namespace:           argo
ServiceAccount:      default
Status:              Succeeded
Created:             Mon Jul 15 01:28:25 +0900 (27 seconds ago)
Started:             Mon Jul 15 01:28:25 +0900 (27 seconds ago)
Finished:            Mon Jul 15 01:28:48 +0900 (4 seconds ago)
Duration:            23 seconds

STEP                                                                              PODNAME                                   DURATION  MESSAGE
 ✔ workflow-template-steps-dlqqb (hello-hello-hello)                                                                
 ├---✔ hello1 (workflow-template-whalesay-template/whalesay-template)             workflow-template-steps-dlqqb-66550121    6s
 └-·-✔ hello2a (workflow-template-inner-steps/inner-steps)                                                          
   | ├---✔ inner-hello1 (workflow-template-whalesay-template/whalesay-template)   workflow-template-steps-dlqqb-1238316130  6s
   | └-·-✔ inner-hello2a (workflow-template-whalesay-template/whalesay-template)  workflow-template-steps-dlqqb-2511940447  7s
   |   └-✔ inner-hello2b (whalesay-template)                                      workflow-template-steps-dlqqb-2528718066  10s
   └-✔ hello2b (workflow-template-whalesay-template/whalesay-template)            workflow-template-steps-dlqqb-769280991   9s

$ ./dist/argo get -n argo workflow-template-dag-diamond-v9m47
Name:                workflow-template-dag-diamond-v9m47
Namespace:           argo
ServiceAccount:      default
Status:              Succeeded
Created:             Mon Jul 15 01:32:59 +0900 (43 seconds ago)
Started:             Mon Jul 15 01:32:59 +0900 (43 seconds ago)
Finished:            Mon Jul 15 01:33:39 +0900 (3 seconds ago)
Duration:            40 seconds

STEP                                                                    PODNAME                                         DURATION  MESSAGE
 ✔ workflow-template-dag-diamond-v9m47 (diamond)                                                                    
 ├-✔ A (workflow-template-whalesay-template/whalesay-template)          workflow-template-dag-diamond-v9m47-3420111096  5s
 ├-✔ B (workflow-template-whalesay-template/whalesay-template)          workflow-template-dag-diamond-v9m47-3470443953  10s
 ├-✔ C (workflow-template-inner-dag/inner-diamond)                                                                  
 | ├-✔ inner-A (workflow-template-whalesay-template/whalesay-template)  workflow-template-dag-diamond-v9m47-1233185554  7s
 | ├-✔ inner-B (whalesay-template)                                      workflow-template-dag-diamond-v9m47-1216407935  9s
 | ├-✔ inner-C (whalesay-template)                                      workflow-template-dag-diamond-v9m47-1199630316  6s
 | └-✔ inner-D (workflow-template-whalesay-template/whalesay-template)  workflow-template-dag-diamond-v9m47-1182852697  6s
 └-✔ D (workflow-template-whalesay-template/whalesay-template)          workflow-template-dag-diamond-v9m47-3503999191  6s
```

dtaniwaki commented:
I updated the steps and DAG example so we can check how nested steps/DAG work.

```
$ ./dist/argo get -n argo workflow-template-steps-dlqqb
Name:                workflow-template-steps-dlqqb
Namespace:           argo
ServiceAccount:      default
Status:              Succeeded
Created:             Mon Jul 15 01:28:25 +0900 (27 seconds ago)
Started:             Mon Jul 15 01:28:25 +0900 (27 seconds ago)
Finished:            Mon Jul 15 01:28:48 +0900 (4 seconds ago)
Duration:            23 seconds

STEP                                                                              PODNAME                                   DURATION  MESSAGE
 ✔ workflow-template-steps-dlqqb (hello-hello-hello)                                                                
 ├---✔ hello1 (workflow-template-whalesay-template/whalesay-template)             workflow-template-steps-dlqqb-66550121    6s
 └-·-✔ hello2a (workflow-template-inner-steps/inner-steps)                                                          
   | ├---✔ inner-hello1 (workflow-template-whalesay-template/whalesay-template)   workflow-template-steps-dlqqb-1238316130  6s
   | └-·-✔ inner-hello2a (workflow-template-whalesay-template/whalesay-template)  workflow-template-steps-dlqqb-2511940447  7s
   |   └-✔ inner-hello2b (whalesay-template)                                      workflow-template-steps-dlqqb-2528718066  10s
   └-✔ hello2b (workflow-template-whalesay-template/whalesay-template)            workflow-template-steps-dlqqb-769280991   9s

$ ./dist/argo get -n argo workflow-template-dag-diamond-v9m47
Name:                workflow-template-dag-diamond-v9m47
Namespace:           argo
ServiceAccount:      default
Status:              Succeeded
Created:             Mon Jul 15 01:32:59 +0900 (43 seconds ago)
Started:             Mon Jul 15 01:32:59 +0900 (43 seconds ago)
Finished:            Mon Jul 15 01:33:39 +0900 (3 seconds ago)
Duration:            40 seconds

STEP                                                                    PODNAME                                         DURATION  MESSAGE
 ✔ workflow-template-dag-diamond-v9m47 (diamond)                                                                    
 ├-✔ A (workflow-template-whalesay-template/whalesay-template)          workflow-template-dag-diamond-v9m47-3420111096  5s
 ├-✔ B (workflow-template-whalesay-template/whalesay-template)          workflow-template-dag-diamond-v9m47-3470443953  10s
 ├-✔ C (workflow-template-inner-dag/inner-diamond)                                                                  
 | ├-✔ inner-A (workflow-template-whalesay-template/whalesay-template)  workflow-template-dag-diamond-v9m47-1233185554  7s
 | ├-✔ inner-B (whalesay-template)                                      workflow-template-dag-diamond-v9m47-1216407935  9s
 | ├-✔ inner-C (whalesay-template)                                      workflow-template-dag-diamond-v9m47-1199630316  6s
 | └-✔ inner-D (workflow-template-whalesay-template/whalesay-template)  workflow-template-dag-diamond-v9m47-1182852697  6s
 └-✔ D (workflow-template-whalesay-template/whalesay-template)          workflow-template-dag-diamond-v9m47-3503999191  6s
```

mark9white commented:
Awesome work on this @dtaniwaki 

Once you've created a WorkflowTemplate is there a way to update it without deleting it first?

dtaniwaki commented:
> Once you've created a WorkflowTemplate is there a way to update it without deleting it first?

@mark9white Updating existing `WorkflowTemplate` is very dangerous because it affects the result of referring workflows. I think this is the discussion point with everyone. Do you expect WorkflowTemplates immutable or not?

dtaniwaki commented:
One idea I came up is we can record template resolution in node status, then completed workflows don't care template modifications.

mark9white commented:
@dtaniwaki That sounds a great idea ...

kbruner commented:
Could they be versioned to address the issue with updates breaking running workflows?

inc0 commented:
Or copied to workflow during workflow execution - so if tpl would change later, it won't affect existing workflows, which makes sense as they're already done.

sarabala1979 commented on cmd/argo/commands/common.go:
```diff
@@ -94,8 +95,8 @@ func InitWorkflowClient(ns ...string) v1alpha1.WorkflowInterface {
 	}
-	wfcs := wfclientset.NewForConfigOrDie(restConfig)
-	wfClient = wfcs.ArgoprojV1alpha1().Workflows(namespace)
+	wfClientset = wfclientset.NewForConfigOrDie(restConfig)
```
Please change the import reference name. it is confusing.

frankfarzan commented:
@dtaniwaki A usability question looking at the example above:

Does it make sense for `templateRef.template` to be an optional field with reasonable defaults to avoid repetition/verbosity? I suspect that in practice, either a WorkflowTemplate is used for reusing a single template (e.g. workflow-template-whalesay-template) OR it contains multiple templates (e.g. workflow-template-nested-template) and the the default can be specified in WorkflowTemplate itself instead of by each reference.

dtaniwaki commented:
@frankfarzan I agree that we want to avoid repetition and verbosity, but I'm not sure what reasonable defaults are. What do you think of having a field which specifies the default template in WorkflowTemplate? I haven't come up with a good name for the field yet, though.

frankfarzan commented:
I see parallels between this field and `spec.entrypoint` in Workflow, so calling this `entrypoint` may be intuitive to users? We can also make the name more descriptive like `defaultTemplate` or something along those lines.

dtaniwaki commented:
@frankfarzan Thanks for your idea about the field name. I think it's better to make that improvement separately after the basic function of this PR gets merged to avoid extra complexity for reviewers.

jessesuen commented on manifests/cluster-install/workflow-controller-rbac/workflow-controller-clusterrole.yaml:
```diff
@@ -36,10 +36,12 @@ rules:
   - workflows
   - workflows/finalizers
+  - workflowtemplates
+  - workflowtemplates/finalizers
```
I think instead of adding to existing rule, it will be better to create a new rule entry so that the controller only has `get`, `list`, and `watch` on templates.

dtaniwaki commented on cmd/argo/commands/template/submit.go:
```diff
@@ -0,0 +1,122 @@
+	return command
+}
+
+func SubmitWorkflowTemplates(filePaths []string, cliOpts *cliSubmitOpts) {
```
Make sense. I'll rename it.

dtaniwaki commented:
@jessesuen I fixed the code based on your review and also added the lint command.

By the way, do you have any idea the ideal behavior of updating templates?
https://github.com/argoproj/argo/pull/1312#issuecomment-511662757
https://github.com/argoproj/argo/pull/1312#issuecomment-511665396

I think we can record resolved template contents in node status so we won't lose what was executed even after updating or deleting templates. We may need to wait for [the Workflow persistence](https://github.com/argoproj/argo/issues/745#issuecomment-485045320) to store massively expanded contents in node status.

jessesuen commented:
> By the way, do you have any idea the ideal behavior of updating templates?
> #1312 (comment)
> #1312 (comment)

> One idea I came up is we can record template resolution in node status, then we can keep executed workflow step information, and it's possible to draw it in UI, even if referred templates got modified or deleted later.

I had the same thought about storing resolved templates in the status upon submission. So when workflow initially starts, we would resolve all templates, store it in status, which would allow templates objects to change even during in-flight workflows. 

I think we could that as a separate PR, I want to get this in faster.

jessesuen commented:
Minor comment about `processAggregateNodeOutputs` and resolve types.go conflict, and it should be good to go.

dtaniwaki commented on workflow/templateresolution/context.go:
```diff
@@ -0,0 +1,144 @@
+// resolveTemplateImpl digs into referenes and returns a merged template.
+func (ctx *Context) resolveTemplateImpl(tmplHolder wfv1.TemplateHolder, depth int) (*Context, *wfv1.Template, error) {
+	// Avoid infinite referenes
+	if depth > maxResolveDepth {
```
But I found a typo the comment above. lol
I'm fixing it.

dtaniwaki commented on workflow/controller/operator.go:
```diff
@@ -1485,9 +1513,9 @@ func (n loopNodes) Swap(i, j int) {
 // processAggregateNodeOutputs adds the aggregated outputs of a withItems/withParam template as a
 // parameter in the form of a JSON list
-func (woc *wfOperationCtx) processAggregateNodeOutputs(templateName string, scope *wfScope, prefix string, childNodes []wfv1.NodeStatus) {
+func (woc *wfOperationCtx) processAggregateNodeOutputs(tmpl *wfv1.Template, scope *wfScope, prefix string, childNodes []wfv1.NodeStatus) error {
```
I added the return error value because new logic there needed at the time of the implementation but changed the logic later but forgot to remove the return value. I'll remove it.

dtaniwaki replied:
However, the original code ignores an error from `Marshal` which is not good. I'll keep the return value but change it to handle the error.

jessesuen commented:
Filed https://github.com/argoproj/argo/issues/1525 to resolve workflow templates into status.

dtaniwaki commented:
#1462 significantly conflicts with this PR. I'm resolving it...

dtaniwaki commented:
@jessesuen I resolved the conflicts and found a test failure. Please check the commits just in case.

hnykda commented:
Is somewhere a link to the documentation for this or is it just a wishful thinking? :grin:

pumicerD commented:
When defining multiple `volumeMounts` in a template linting / submitting fails with :

`templates.foo templates.foo.container.volumeMounts[1].mountPath 'placeholder' already mounted in container.volumeMounts.placeholder`

Linting succeeds when I comment out the second `volumeMount`.
Seems like all volume mounts are being mapped to the same placeholder ?

Ark-kun commented:
This change is very weird and confusing from the design perspective.

Argo has two major types - `Template` and `Task`.
`Task` ~= `template` + `arguments`

It would have been natural to add `templateRef` to `Task` so that you can have 
`Task` ~= `templateRef` + `arguments`

Instead this PR does something really weird by adding `templateRef` to `Teamplate`
Now template can reference template which can reference template etc. Meanwhile each template in that infinite stack has separate inputs, outputs, dag, etc. This just seems very wrong from API or DB architecture perspective.

I also see that this PR has added `arguments` to `Template`.
So, previously `Task` ~= `template` + `arguments`.
Now `Template` also ~= `Template` + `arguments`
What does that mean?

Ark-kun commented on api/openapi-spec/swagger.json:
```diff
@@ -839,6 +843,10 @@
           "description": "Location in which all files related to the step will be stored (logs, artifacts, etc...). Can be overridden by individual items in Outputs. If omitted, will use the default artifact repository location configured in the controller, appended with the \u003cworkflowname\u003e/\u003cnodename\u003e in the key.",
           "$ref": "#/definitions/io.argoproj.workflow.v1alpha1.ArtifactLocation"
         },
+        "arguments": {
```
Are you sure this belongs to the `Template` struct? AFAIK it should be in the `DagTask` struct where it already exists.

Ark-kun commented on api/openapi-spec/swagger.json:
```diff
@@ -218,6 +218,10 @@
           "description": "Name of template to execute",
           "type": "string"
         },
+        "templateRef": {
```
Are you sure this belongs to the `Template` struct? AFAIK it should be in the `DagTask` as alongside the `template` attribute.

Ark-kun commented on pkg/apis/workflow/v1alpha1/workflow_types.go:
```diff
@@ -182,6 +198,15 @@ type Template struct {
+	Template string `json:"template,omitempty"`
+
+	// Arguments hold arguments to the template.
+	Arguments Arguments `json:"arguments,omitempty"`
```
Are you sure this belongs to the `Template` struct? AFAIK it should be in the `DagTask` struct where it already exists.

P.S> The omitempty does not work in practice since some code initializes it to empty array.

Ark-kun commented on pkg/apis/workflow/v1alpha1/workflow_types.go:
```diff
@@ -182,6 +198,15 @@ type Template struct {
 	Name string `json:"name"`
 
+	// Template is the name of the template which is used as the base of this template.
+	Template string `json:"template,omitempty"`
```
Are you sure this belongs to the `Template` struct? AFAIK it should be in the `DagTask` struct where it already exists.

Ark-kun commented on pkg/apis/workflow/v1alpha1/workflow_types.go:
```diff
@@ -182,6 +198,15 @@ type Template struct {
+	Arguments Arguments `json:"arguments,omitempty"`
+
+	// TemplateRef is the reference to the template resource which is used as the base of this template.
+	TemplateRef *TemplateRef `json:"templateRef,omitempty"`
```
Are you sure this belongs to the `Template` struct? AFAIK it should be in the `DagTask` struct alongside the `Template` field.

Ark-kun commented:
This could have been a small nice design...

Just
```yaml
apiVersion: argoproj.io/v1alpha1
kind: Workflow
spec:
  entrypoint: whalesay
  templates:
  - name: whalesay
    templateRef:
      url: https://..... # Can point to any template YAML - container, DAG, etc
    arguments:
      parameters:
      - name: message
        value: "hello"
```
Similar to https://github.com/kubeflow/pipelines/blob/fd6c756dd2a510581e52557cee881ec0c2cbc445/sdk/python/tests/components/test_data/retail_product_stockout_prediction_pipeline.component.yaml#L40

Ark-kun commented on pkg/apis/workflow/register.go:
```diff
@@ -2,10 +2,15 @@ package workflow
-	ShortName string = "wf"
-	FullName  string = Plural + "." + Group
+	Group                     string = "argoproj.io"
+	WorkflowKind              string = "Workflow"
```
Is this a breaking change?
AFAIK, Go package versioning rules strongly discourage making breaking changes. Especially in point releases.

https://golang.org/doc/faq#get_version

>the import compatibility rule: “If an old package and a new package have the same import path, the new package must be backwards compatible with the old package.”

>don't remove exported names, encourage tagged composite literals, and so on. If different functionality is required, add a new name instead of changing an old one. If a complete break is required, create a new package with a new import path.

Can you tell us a bit more about the emergency circumstances that required renaming these variables?
